### PR TITLE
Slurmrestd Charm Unit Tests

### PR DESCRIPTION
**Pull request by dvdgomez**
## Summary of Change

Added unit tests covering the slurmrestd charm functionality.

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -12,9 +12,109 @@ def setUp(self) -> None:
         self.harness.begin()
         self.maxDiff = None
```
Comment (NucciTheBoss) [inline tests/unit/test_charm.py]: Similar to the `slurmd` unit tests, I do not believe that you need `self.maxDiff`.

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -12,9 +12,109 @@ def setUp(self) -> None:
...
+    @patch("slurm_ops_manager.SlurmManager.install")
+    @patch("pathlib.Path.read_text")
+    @patch("ops.model.Unit.set_workload_version")
+    @patch("slurm_ops_manager.SlurmManager.start_munged", lambda _: True)
+    @patch("ops.framework.EventBase.defer")
+    def test_install_success(self, defer, *_):
+        self.harness.charm.on.install.emit()
+        self.assertTrue(self.harness.charm._stored.slurm_installed)
+        self.assertEqual(self.harness.charm.unit.status, ActiveStatus("slurmrestd available"))
+        defer.assert_not_called()
```
Comment (NucciTheBoss) [inline tests/unit/test_charm.py]: Similar to the `slurmd` charm, do we really want to mock installing `slurmrestd`. This method success is hinged on `SlurmManager.install` working for `slurmrestd`. I think you should make sure that the install method at least works.

Also, not sure if we want to be patching out that the version file is missing. I think how version is set at install is a major functional issue that needs to be address when we start refactoring the slurm charms.

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -26,3 +126,22 @@ def test_update_status_success(self, *_):
 
         self.harness.charm.on.update_status.emit()
         self.assertEqual(self.harness.charm.unit.status, ActiveStatus("slurmrestd available"))
+
+    @patch("pathlib.Path.read_text")
+    @patch("ops.model.Unit.set_workload_version")
+    def test_upgrade_fail(self, *_):
+        self.harness.charm.on.upgrade_charm.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus("Error installing slurmrestd"))
```
Comment (NucciTheBoss) [inline tests/unit/test_charm.py]: Might not change anything that much, but you can have the patched `pathlib.Path.read_text` return a string value like "v1.0.0" or something similar.

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -26,3 +126,22 @@ def test_update_status_success(self, *_):
...
+    @patch("slurm_ops_manager.SlurmManager.needs_reboot", new_callable=PropertyMock(return_value=False))
+    @patch("interface_slurmrestd.SlurmrestdRequires.is_joined", new_callable=PropertyMock(return_value=True))
+    @patch("interface_slurmrestd.SlurmrestdRequires.get_stored_munge_key", lambda _: True)
+    @patch("interface_slurmrestd.SlurmrestdRequires.get_stored_jwt_rsa", lambda _: True)
+    @patch("interface_slurmrestd.SlurmrestdRequires.get_stored_slurm_config", lambda _: True)
+    def test_upgrade_success(self, *_):
+        self.harness.charm._stored.slurm_installed = True
+
+        self.harness.charm.on.upgrade_charm.emit()
+        self.assertEqual(self.harness.charm.unit.status, ActiveStatus("slurmrestd available"))
```
Comment (NucciTheBoss) [inline tests/unit/test_charm.py]: Same with testing install success, do we want to be patching out the fact that the version file is missing? Might lull another contributor into a false sense of security that their copy of the charm is working as expected. I think you should not patch the version file and let the test fail. I believe that one thing we should be focused on while making our improvements to the charms is figuring out how we can get rid of this version file.

**Review by NucciTheBoss — CHANGES_REQUESTED** [review]
Left a few comments. Looks good overall :smile:

File: tests/unit/test_charm.py (hunk 4 above)
Comment (dvdgomez) [reply tests/unit/test_charm.py]: I agree, no longer patching out version file.

File: tests/unit/test_charm.py (hunk 1 above)
Comment (dvdgomez) [reply tests/unit/test_charm.py]: Removed

File: tests/unit/test_charm.py (hunk 2 above)
Comment (dvdgomez) [reply tests/unit/test_charm.py]: No longer patching out the version file

**Review by NucciTheBoss — APPROVED** [review]
Excellent :rocket:! Thanks for implementing my feedback. I am going to go ahead and merge this pull request.

**Pull request merged**